webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java

```diff
@@ -21,28 +25,40 @@
 
     private JdbcTemplate jdbcTemplate;
     private MeterRegistry meterRegistry;
+    private Map<PendingJob, AtomicLong> queueSizes;
 
     public QuartzPendingJobsReportingTask(@Autowired DataSource dataSource,
                                           @Autowired MeterRegistry meterRegistry) {
         this.meterRegistry = meterRegistry;
         this.jdbcTemplate = new JdbcTemplate(dataSource);
+        this.queueSizes = new ConcurrentHashMap<>();
     }
 
     @Scheduled(fixedRateString = "${l10n.management.metrics.quartz.sql-queue-monitoring.execution-rate}")
     public void reportPendingJobs() {
-        fetchResults().forEach(this::reportResults);
+        // First we remove all of our previously stored results (notice line 52). If a job clears from the queue
+        // we would still report it as having pending jobs as it'll retrieve its size from our queueSizes map.
+        queueSizes.clear();
+        fetchResults().forEach(this::registerResults);
     }
 
-    private void reportResults(PendingJob pendingJob){
-        meterRegistry.gauge("quartz.pending.jobs",
-                            Tags.of("jobClass", pendingJob.jobClass, "jobGroup", pendingJob.jobGroup),
-                            pendingJob.count);
+    private void registerResults(PendingJob pendingJob) {
+
+        // For each pending job we store the value we got from the db regardless of its current value in the map
+        queueSizes.put(pendingJob, new AtomicLong(pendingJob.count));
```

Comment:
why don't we persist the gauge and then update their value? https://micrometer.io/docs/concepts#_manually_incrementingdecrementing_a_gauge

Reply:
We'd still need to maintain a list of previously generated values. At the end of the day is the same issue we're solving here with a map/cache along with a supplier being passed to the gauge builder.

Reply:
I'm not sure to understand why we need previously generate values? we just need to assign the new values to the gauges, no?

Reply:
Let me use an example:

- At t0, we call `fetchResults()` and we get two results, job1 having 3 pending tasks, job2 having 1 pending task. We report those values to the registry.

- At t1, we get one result, job1 having 2 pending tasks. Pending tasks for job2 are cleared, therefore our `List<PendingJob>` won't have any items from job2 within its results, so we won't update its count to 0 and rather we'll keep it as 1 (the value reported at t0). But since we do `queueSize.clear()` and the we use `() -> queueSizes.getOrDefault(pendingJob, new AtomicLong(0))` as the supplier that will generate values for that gauge, we ensure the value goes back to 0 for that specific gauge (the one we use for job2).

With the direct `Number` reference we lose track of gauges that yielded results previously and in the current measurement, do not return any result.

Reply:
I'm thinking of pseudo code similar to this:

```
Map gauges = ....
results = fetchResults()
gaugesWithoutResult = ...
gaugeWithResult = ...
resultWithoutGauge = ....
for (gauge :  gaugesWIthoutResult) gauge.set(0)
for (gauge :  gaugesWIthResult) gauge.set(result.getValue())
for (result :  resultWithoutGauge) createGauge()
```

gaugesWithoutResult and gaugeWithResult could also be one and just have a getValue() that either return 0 if there is no entry in `results` or the actual value from `results`